src/GafferScene/BranchCreator.cpp

```diff
@@ -321,65 +381,61 @@ IECore::ConstInternedStringVectorDataPtr BranchCreator::computeChildNames( const
 
 void BranchCreator::hashSetNames( const Gaffer::Context *context, const ScenePlug *parent, IECore::MurmurHash &h ) const
 {
-	ScenePlug::ScenePath parentPath;
-	string parentAsString = parentPlug()->getValue();
-	if( parentAsString.empty() )
-	{
-		// no parent specified so setNames will be unchanged
-		h = inPlug()->setNamesPlug()->hash();
-		return;
-	}
-	ScenePlug::stringToPath( parentAsString, parentPath );
+	ConstPathMatcherDataPtr parentPathsData = parentPathsPlug()->getValue();
+	const PathMatcher &parentPaths = parentPathsData->readable();
 
-	MurmurHash branchSetNamesHash;
-	hashBranchSetNames( parentPath, context, branchSetNamesHash );
-	if( branchSetNamesHash == MurmurHash() )
+	if( parentPaths.isEmpty() )
 	{
 		h = inPlug()->setNamesPlug()->hash();
 		return;
 	}
 
-	SceneProcessor::hashSetNames( context, parent, h );
+	FilteredSceneProcessor::hashSetNames( context, parent, h );
 	inPlug()->setNamesPlug()->hash( h );
-	h.append( branchSetNamesHash );
+
+	/// \todo In practice all BranchCreators in Gaffer have set names which do not vary
```

Comment:
@danieldresser-ie, any thoughts on this? I'm thinking I'd like to address it one way or another - either by requiring all branches to have the same names, or by providing a `virtual branchSetNamesAreConstant()` method to enable the optimisation. Any preference?

Reply:
@danieldresser-ie, any comments? Now Murray is back, I'd like to get this merged for production-scale testing, but I don't want to do it without addressing this.

Reply:
Hmm, I guess it wouldn't be that bad to require all branches to have the same names, with a comment about how to add branchSetNamesAreConstant() if needed in the future ...

I guess to me, the question is whether we have any actual plans for nodes that would use branchSetNamesAreConstant().

I'm contemplating a use case where you wanted to use a context variable to put different stuff at a bunch of different locations in a hierarchy ... though, wait, based on how we approach instancers now, we would require the user to first use a CollectScenes to create a set of objects, and then some weird branchCreator thing to place them into the hierarchy.  So even if we did need to do this, the set accumulation would be done by CollectScenes, not the branchCreator derived node.

So I guess I can't yet see a need for branchSetNamesAreConstant() - but it probably would be good to mention it in a comment in case the matching set names requirement becomes a problem somewhere down the line.

Reply:
> I guess to me, the question is whether we have any actual plans for nodes that would use branchSetNamesAreConstant().

I don't have any plans, but adding the virtual function in the future would require a major version, which might not be convenient if/when it does come up. Let's discuss in the meeting today.

Reply:
Hmm, we didn't talk about this.  I don't have strong feelings about it really.  I guess the question is just whether we can brainstorm up a plausible use case in the future for branchSetNamesAreConstant()?

Reply:
The API generally is about allowing a subclass to make any branches it wants, so it seems reasonable to allow arbitrary set names per branch even if we don't have a requirement just now. I've already written the code for the complex case, and you've already reviewed it, so slotting in the optimised path triggered by `branchSetNamesAreConstant()` is possibly the easiest course of action too.

Reply:
Sounds reasonable.

Reply:
Added in 87b52eb